ZenPacks/zenoss/ZenPackLib/lib/helpers/utils.py

```diff
@@ -0,0 +1,248 @@
+    data = yaml.load(orig_yaml)
+    # optimized output
+    optimized_yaml = get_optimized_yaml(data)
+    # new load
+    valid = compare_zenpackspecs(orig_yaml, optimized_yaml)
+    if not valid:
+        LOG.warn('OPTIMIZATION FAILED VERIFICATION!  Please review optimized YAML prior to use ')
+    return optimized_yaml
+
+
+def compare_zenpackspecs(orig_yaml, new_yaml):
+    """report whether different YAML documents are identical"""
+    # now load both yaml files and create ZenPackSpec configs
+    # apply log verbosity settings
+    Loader.QUIET= False
+    Loader.LEVEL = 0
+    orig = yaml.load(orig_yaml, Loader=Loader)
+    new = yaml.load(new_yaml, Loader=Loader)
+    orig.create()
+    new.create()
+    # SpecParams should be equal
+    if orig != new:
+        LOG.warn('ZenPackSpec mismatch between original and new')
+        dict_compare(orig.__dict__, new.__dict__)
+    # now dump new specs back out to yaml
+    orig_dump = yaml.dump(orig.specparams, Dumper=Dumper)
+    new_dump = yaml.dump(new.specparams, Dumper=Dumper)
+    # load raw yaml into Python dictionaries
+    orig_raw_yaml = yaml.load(orig_dump)
+    new_raw_yaml = yaml.load(new_dump)
+    # these should also be equivalent
+    if orig_raw_yaml != new_raw_yaml:
+        LOG.warn('YAML loaded Python dictionary mismatch between original and new')
+        dict_compare(orig_raw_yaml, new_raw_yaml)
+        return False
+    return True
+
+def dict_compare(d1, d2):
+    d1_keys = set(d1.keys())
+    d2_keys = set(d2.keys())
+    intersect_keys = d1_keys.intersection(d2_keys)
+    added = d1_keys - d2_keys
+    removed = d2_keys - d1_keys
+    modified = {o : (d1[o], d2[o]) for o in intersect_keys if d1[o] != d2[o]}
+    same = set(o for o in intersect_keys if d1[o] == d2[o])
+    LOG.warn('ADDED: {}'.format(added))
+    LOG.warn('REMOVED: {}'.format(removed))
+    LOG.warn('MODIFIED: {}'.format(modified))
+
+
+def get_optimized_yaml(data):
+    """
+        return optimized YAML representing ZenPackSpec
+    """
+    # set DEFAULTS throughout
+    descend_defaults(data)
+    # sort
+    ordered = sort_yaml_data(data)
+    # optimized output
+    return yaml.dump(ordered,  default_flow_style=False, Dumper=Dumper).replace('!!map','')
+
+
+def descend_defaults(input):
+    """
+        descend YAML dictionary and optimize DEFAULTS
+    """
+    for k, v in input.items():
+        if not isinstance(v, dict):
+            continue
+        if k == 'DEFAULTS': 
+            continue
+        if k in ['classes', 'properties', 'thresholds', 'datasources', 'datapoints', 'graphs', 'relationships','graphpoints']:
+            if len(v.keys()) > 1:
+                set_defaults(v)
+        descend_defaults(v)
+
+
+def set_defaults(input):
+    """
+        if dict has multiple entries, look for identical key/value 
+        pairs and move them to DEFAULTS
+    """
+    defaults = {}
+    # get list of potential defaults
+    for name, params in input.items():
```

Comment:
the "set_defaults" never gets called unless the "input" param is a dict, so it shouldn't be a problem here